lib/icinga/dependency.ti

```diff
@@ -50,13 +50,13 @@ class Dependency : CustomVarObject < DependencyNameComposer
 		}}}
 	};
 
-	[config, required, navigation(parent_host)] name(Host) parent_host_name {
+	[config, no_user_modify, required, navigation(parent_host)] name(Host) parent_host_name {
 		navigate {{{
 			return Host::GetByName(GetParentHostName());
 		}}}
 	};
 
-	[config, navigation(parent_service)] String parent_service_name {
+	[config, no_user_modify, navigation(parent_service)] String parent_service_name {
```

Comment:
These attributes don't affect the object name, so from that point of view, changing them is fine.

Please try to figure out whether it's safe to change them at runtime, keep or revert this change accordingly and give the reasoning for it.

Reply:
Do you consider this as acceptable?

```bash
object Dependency "dependency" {
    parent_host_name = NodeName
    parent_service_name = "random1"

    child_host_name = "icinga-0"
    child_service_name = "random0"

    disable_checks = true
}
```

```
[2022-09-09 14:36:11 +0200] notice/Dependency: Dependency 'icinga-0!random0!dependency' passed: Parent service 'yonass-mbp.int.netways.de!random1' matches state filter.
[2022-09-09 14:36:12 +0200] notice/Dependency: Dependency 'icinga-0!random0!dependency' passed: Parent service 'yonass-mbp.int.netways.de!random1' matches state filter.
```

Now, modify `parent_service_name`: 
```bash
# curl -k -s -S -i -u root:icinga -H 'Accept: application/json' \
 -X POST 'https://localhost:5665/v1/objects/dependencies/icinga-0!random0!dependency' \
 -d '{ "attrs": { "parent_service_name": "random0" }, "pretty": true }'

{
    "results": [
        {
            "code": 200,
            "name": "icinga-0!random0!dependency",
            "status": "Attributes updated.",
            "type": "Dependency"
        }
    ]
}
```

However, the dependency graph is still not updated.
```bash
[2022-09-09 14:36:23 +0200] notice/Dependency: Dependency 'icinga-0!random0!dependency' passed: Parent service 'yonass-mbp.int.netways.de!random1' matches state filter.
[2022-09-09 14:36:24 +0200] notice/Dependency: Dependency 'icinga-0!random0!dependency' passed: Parent service 'yonass-mbp.int.netways.de!random1' matches state filter.
```

Reply:
I think this is because the dependencies are only updated when the `OnAllConfigLoaded` event is fired, which does not happen when modifying an attribute. The changed attributes are also applied, but at that point it is too late to update the parent-child dependency graph.

https://github.com/Icinga/icinga2/blob/14e4f6b9213913e00b9a4612bb11fc13882b8b10/lib/icinga/dependency.cpp#L78-L91

Reply:
> Do you think this is acceptable?

This looks broken. So then the reasoning for not allowing updates is "changes the object name which is not allowed" for some and "actual dependencies aren't updated" for others. (Though for the latter it could just be a valid option to fix this, but maybe not right now so not allowing the update rather then accepting it and then showing broken behavior is probably the better option in the meantime.)

> Even after restarting Icinga2, the modified `parent_service_name` doesn't get applied.

> ```                "parent_service_name": "random0",```

Looks fine to me?

Reply:
> > Even after restarting Icinga2, the modified `parent_service_name` doesn't get applied.
> 
> > `                "parent_service_name": "random0",`
> 
> Looks fine to me?

I was referring to the dependency!

Reply:
> "actual dependencies aren't updated"

This is exactly what’s happening here.